text/0000-signals-in-lwc-templates.md

```diff
@@ -0,0 +1,196 @@
+    <p>{count.value}</p>
+</template>
+```
+
+### Compiler Enhancements
+
+The LWC compiler will be updated to recognize when a Signal's `.value` is used within a template. Instead of accessing the property directly at runtime, it will generate code to subscribe to the Signal's changes during component's template initialization and unsubscribe upon component's template disposal.
+
+### Store Protocol
+
+A store must implement the following protocol to be compatible with the new reactivity model:
+
+```ts
+class Store {
+    get value(): any { /* return current value */ }
+    subscribe((newValue: any) => void): () => void { /* subscribe to changes */ }
+}
+```
+
+__NOTE:__ _We can work the details of this protocol. It could be more simple, or hopefully similar to what is out there._
+
+### Integration with Components
+
+Components should treat signals as internal reactive state mechanisms. While a signal can technically be passed down to child components, it's recommended to pass only the necessary data, typically the `.value` of the signal, to keep the child components decoupled from the parent's state management strategy.
+
+Here’s how a parent component would pass the current value of a signal to a child component, which expects a number:
+
+```javascript
+// ParentComponent.js
+import { signal } from 'some/signals';
+
+export default class ParentComponent extends LightningElement {
+    count = signal(0);
+}
+```
+
+```html
+<!-- ParentComponent's template -->
+<template>
+    <!-- Pass the value of the signal, not the signal itself -->
+    <child-component count={count.value}></child-component>
+</template>
+```
+
+The child component simply receives a number as a property and uses it as such, unaware of the signal mechanism in the parent:
+
+```javascript
+export default class ChildComponent extends LightningElement {
+    @api count;
+}
+```
+
+This ensures that child components remain agnostic of the parent's state management implementation, promoting better component encapsulation and reuse.
+
+In terms of optimizations, the compiler can detect when a signal is passed down to a child component because of the `.value` in the template interpolation, and generate code to subscribe to the signal's changes. This would allow the parent component to update the child's public property `count` without a full re-render, leading to better performance. On the receiving end, because the child component's `count` property is not a signal, but a value, it will react to changes as usual, without any special handling.
+
+### Object Reactivity
+
+Objects can be wrapped in Signals to make them reactive. However, changes to nested properties must be handled by replacing the entire object:
+
+```javascript
+import { signal } from 'signals';
+
+export default class ExampleComponent extends LightningElement {
+    position = signal({ x: 0, y: 0 });
+
+    move(x, y) {
+        this.position.value = { ...this.position.value, x, y };
+    }
```

Comment:
What if you allow developers to configure this behavior?

```js
position = signal({x:0, y:0}) //has to change the whole object

position = signal({x:0,y:0}, { deep: true }) //deep comparison will be made which enables signaling based on prop changes
```

Reply:
@AllanOricil all that is possible, but not part of this RFC. Remember, this is about the template, and how to use signals in the template, how to create and update signals, that's a different RFC.

As for the specific syntax in the template, you will probably need to do: `<div>{{position.value.x}}</div>`. The `.value` is the signal to listen for changes at runtime.